onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc

```diff
@@ -287,187 +323,260 @@
-          data[i].C = y_data + helper.OutputOffsets()[i];
-          data[i].ldc = N;
-        }
-        MlasSQNBitGemmBatch(M, N, K, batch_count, nbits_, block_size_, compute_type, data.data(), workspace.get(),
-                            thread_pool);
-        return Status::OK();
-      }
+    if (compute_type_ == CompInt8) {
+      data[i].QuantBDataWorkspace = packed_b_.get();
     }
+#endif
+    data[i].PackedQuantBData = static_cast<std::byte*>(packed_b_.get());
+    data[i].QuantBScale = tmp_scales_data_ptr.get();
+    data[i].QuantBZeroPoint = zero_points_data;
+    data[i].Bias = bias_data != nullptr ? &bias_data_v[0] : nullptr;
+    data[i].C = &C_v[0] + helper.OutputOffsets()[i];
+    data[i].ldc = N;
   }
+  MlasSQNBitGemmBatch(M, N, K, batch_count, nbits_, block_size_, compute_type_, data.data(), workspace.get(),
+                      thread_pool);
+  MlasConvertFloatToHalfBuffer(&C_v[0], y_data, C_v.size());
+  return Status::OK();
+}
 
-#endif  // !defined(ORT_NEURAL_SPEED)
+template <>
+Status MatMulNBits<float>::ComputeBUnpacked(const Tensor* a,
+                                            const Tensor* b,
+                                            const Tensor* scales,
+                                            const Tensor* zero_points,
+                                            const Tensor* reorder_idx,
+                                            const Tensor* bias,
+                                            Tensor* y,
+                                            AllocatorPtr& allocator,
+                                            concurrency::ThreadPool* thread_pool,
+                                            const MatMulComputeHelper& helper) const {
+  const auto* a_data = a->Data<float>();
+  const uint8_t* b_data = b->Data<uint8_t>();
+  const auto* scales_data = scales->Data<float>();
+  const auto* zero_points_data = zero_points == nullptr ? nullptr : zero_points->DataRaw();
+  const auto* reorder_idx_data = reorder_idx == nullptr ? nullptr : reorder_idx->Data<int32_t>();
+  auto* y_data = y->MutableData<float>();
 
-  // fallback implementation - dequantize B first and then compute float gemm
+  const size_t batch_count = helper.OutputOffsets().size();
+  const size_t M = static_cast<size_t>(helper.M());
+  const size_t N = static_cast<size_t>(helper.N());
+  const size_t K = static_cast<size_t>(helper.K());
+  const size_t lda = helper.Lda(false);
+  const size_t ldb = helper.Ldb(true);
 
-  const Tensor* scales = ctx->Input<Tensor>(InputIndex::scales);
-  const Tensor* zero_points = ctx->Input<Tensor>(InputIndex::zero_points);
-  const Tensor* reorder_idx = ctx->Input<Tensor>(InputIndex::g_idx);
+  auto tmp_b_data_ptr = IAllocator::MakeUniquePtr<float>(allocator, SafeInt<size_t>(K_) * N_);
 
-  const auto* scales_data = scales->Data<AType>();
-  const float* scales_data_;
-  std::vector<float> scales_data_v;
-  if constexpr (std::is_same<AType, MLFloat16>::value) {
```

Comment:
as you commented previously, could you please use allocator to allocate a temporary buffer to instead of using vector?

Reply:
I remember Liqun memtioned about memory footprint. I guess Liqun chose vector to reduce memory footprint. + @liqunfu could you pls share the reason of using vector instead of allocator here? thanks